py/picca/delta_extraction/astronomical_objects/forest.py

```diff
@@ -122,73 +121,78 @@ def rebin(log_lambda, flux, ivar, transmission_correction, z, wave_solution,
                10**log_lambda_rest_frame_grid[0] - half_pixel_step_rest_frame)
         w1 &= (lambda_ / (1. + z) <
                10**log_lambda_rest_frame_grid[-1] + half_pixel_step_rest_frame)
-        w1 &= (ivar > 0.)
+        # w1 &= (ivar > 0.)
     else:
         raise AstronomicalObjectError("Error in Forest.rebin(). "
                                       "Class variable 'wave_solution' "
                                       "must be either 'lin' or 'log'.")
 
+    if (w1 & (ivar > 0.)).sum() == 0:
+        log_lambda = np.zeros(0)
+        flux = np.zeros(0)
+        ivar = np.zeros(0)
+        transmission_correction = np.zeros(0)
+        mean_snr = 0.0
+        bins = np.zeros(0, dtype=np.int64)
+        rebin_ivar = np.zeros(0)
+        w1 = np.zeros(0, dtype=bool_)
+        w2 = np.zeros(0, dtype=bool_)
+        return (log_lambda, flux, ivar, transmission_correction, mean_snr, bins,
+                rebin_ivar, orig_ivar, w1, w2)
+
     log_lambda = log_lambda[w1]
     flux = flux[w1]
     ivar = ivar[w1]
     transmission_correction = transmission_correction[w1]
-    if w1.sum() == 0:
-        log_lambda = np.zeros(log_lambda.size)
-        flux = np.zeros(log_lambda.size)
-        ivar = np.zeros(log_lambda.size)
-        transmission_correction = np.zeros(log_lambda.size)
-        mean_snr = 0.0
-        bins = np.zeros(log_lambda.size, dtype=np.int64)
-        rebin_ivar = np.zeros(log_lambda.size)
-        w1 = np.zeros(log_lambda.size, dtype=bool_)
-        w2 = np.zeros(log_lambda.size, dtype=bool_)
-        return (log_lambda, flux, ivar, transmission_correction, mean_snr, bins,
-                rebin_ivar, orig_ivar, w1, w2)
 
     bins = find_bins(log_lambda, log_lambda_grid, wave_solution)
+    binned_arr_size = bins.max() + 1
     log_lambda = log_lambda_grid[0] + bins * pixel_step
 
+    # Find non-empty bins
+    bincounts = np.bincount(bins, minlength=binned_arr_size)
+    wnonempty_bins = bincounts != 0
```

Comment:
I can't remember well. I think there are empty bins on the lower end by construction?

Reply:
hm, I guess this could happen depending on how the wavelength grid is chosen in restframe and observed frame. But ideally we should at least prevent this from happening in the default setup. Maybe I'm ignoring some detail, @iprafols?